File: .agents/skills/revit-transactions/assets/TransactionTemplates.cs
using System;
using Autodesk.Revit.DB;

namespace TransactionTemplates
{
    public static class TransactionHelper
    {
        /// <summary>
        /// Standard Transaction block using the mandatory 'using' statement.
        /// </summary>
        public static void StandardTransactionExample(Document doc)
        {
            using (Transaction tx = new Transaction(doc, "Standard Operation"))
            {
                tx.Start();
                try
                {
                    // Logic to modify the document goes here
                    // ...

                    tx.Commit();
                }
                catch (Exception)
                {
                    tx.RollBack();
                    throw;
                }
            }
        }

        /// <summary>
        /// Safe execution block that checks if the document is already modifiable.
        /// Uses a SubTransaction if a parent transaction exists, otherwise creates a new Transaction.
        /// </summary>
        public static void SafeExecutionExample(Document doc, Action action)
        {
            if (doc.IsModifiable)
            {
                // We are already inside a transaction (e.g. called from another method)
                using (SubTransaction subTx = new SubTransaction(doc))
                {
                    subTx.Start();
                    try
                    {
                        action();
                        subTx.Commit();
                    }
                    catch
                    {
                        subTx.RollBack();
                        throw;
                    }
                }
            }
            else
            {
                // No active transaction exists
                using (Transaction tx = new Transaction(doc, "Safe Operation"))
                {
                    tx.Start();
                    try
                    {
                        action();
                        tx.Commit();
                    }
                    catch
                    {
                        tx.RollBack();
                        throw;
                    }
                }
            }
        }
    }
}

File: .agents/skills/revit-transactions/assets/transaction_templates.py
# pyRevit and IronPython Transaction Templates

import clr
clr.AddReference('RevitAPI')
from Autodesk.Revit.DB import SubTransaction
from pyrevit import revit

# 1. Standard pyRevit Context Manager
# This is the preferred way to execute transactions in pyRevit scripts.
# It automatically starts, commits (or rolls back on exception), and refreshes the UI.
def standard_pyrevit_transaction():
    doc = revit.doc
    
    with revit.Transaction("pyRevit Operation"):
        # Logic to modify the document goes here
        # e.g., element.LookupParameter("Comments").Set("Updated")
        pass


# 2. Nested SubTransaction in Python
# Use this when you are already inside a pyRevit Transaction (or a clean transaction environment)
# and you want to try an operation that might fail, without rolling back the entire parent transaction.
def subtransaction_example():
    doc = revit.doc
    
    with revit.Transaction("Main Operation"):
        # Perform some main operation that we want to keep
        # ...

        # Attempt a risky sub-operation
        sub_tx = SubTransaction(doc)
        sub_tx.Start()
        
        try:
            # Risky logic goes here
            # ...
            sub_tx.Commit()
        except Exception as e:
            # Rollback only the sub-transaction if it fails
            sub_tx.RollBack()
            print("Sub-operation failed, but main operation continues: {}".format(e))
